force-app/main/default/lwc/miscPermissionBasedUI/__tests__/miscPermissionBasedUI.test.js

```diff
@@ -19,53 +19,122 @@ describe('c-misc-permission-based-u-i', () => {
         while (document.body.firstChild) {
             document.body.removeChild(document.body.firstChild);
         }
+        jest.clearAllMocks();
     });
 
-    it('displays the correct UI when custom permission is true', () => {
+    // Helper function to wait until the microtask queue is empty. This is needed for promise
+    // timing when calling imperative Apex.
+    async function flushPromises() {
+        return Promise.resolve();
+    }
+
+    it('displays the correct UI when custom permission is true', async () => {
+        mockPermission.mockReturnValueOnce({
+            ...mockModule,
+            default: true
+        });
+
+        let MiscPermissionBasedUI;
+        jest.isolateModules(() => {
+            MiscPermissionBasedUI = require('../miscPermissionBasedUI').default;
+        });
+
         const element = createElement('c-misc-permission-based-u-i', {
```

Comment:
It is throwing an error here and unable to create the element. Not sure if it has something to do with ES6.
![Screen Shot 2021-08-09 at 12 11 50 PM](https://user-images.githubusercontent.com/31529847/128668628-5035b373-7815-4bbd-a4cb-3597e6de241d.png)

Reply:
Hi filiprafalowicz,

The test cases work if we create the component within the 'isolateModules' call. Here is a sample. Would you please make this change in all the test cases and commit to the PR so that I can merge?

`it('displays the correct UI when custom permission is undefined', async () => {
        mockPermission.mockReturnValueOnce({
            ...mockModule,
            default: undefined
        });
        jest.isolateModules(async () => {
            const element = createElement('c-misc-permission-based-u-i', {
                is: MiscPermissionBasedUI
            });
            document.body.appendChild(element);
            // Wait for any asynchronous DOM updates
            await flushPromises();
            const pEl = element.shadowRoot.querySelector('p');
            expect(pEl.textContent).toBe('The permission set is not assigned');
        });
    });`

Reply:
@satyasekharcvb unfortunately isolateModules doesn't accept async functions. There is an issue in jest open for it: https://github.com/facebook/jest/issues/10428. Also, we need to use require to make it work as isolateModules only isolates the modules that are imported within its scope. If we import MiscPermissionBasedUI on the top of the file, we won't be able to manage different states of custom permission.

Reply:
One thing I can suggest to not duplicate the code in all test methods would be to have something like this:
```
import { createElement } from 'lwc';

const miscPermissionBasedUIModulePath = '../miscPermissionBasedUI';

// Mocking custom permission module
const mockModule = {
    __esModule: true
};

const mockPermission = jest.fn();

jest.mock(
    '@salesforce/customPermission/accessRestrictedUIPermission',
    () => mockPermission(),
    { virtual: true }
);

describe('c-misc-permission-based-u-i', () => {
    afterEach(() => {
        // The jsdom instance is shared across test cases in a single file so reset the DOM
        while (document.body.firstChild) {
            document.body.removeChild(document.body.firstChild);
        }
        jest.clearAllMocks();
    });

    // Helper function to wait until the microtask queue is empty. This is needed for promise
    // timing when calling imperative Apex.
    async function flushPromises() {
        return Promise.resolve();
    }

    function createIsolatedElement(tagName, options, modulePath) {
        let module;

        jest.isolateModules(() => {
            module = require(modulePath).default;
        });

        return createElement(tagName, {
            ...options,
            is: module
        });
    }

    it('displays the correct UI when custom permission is true', async () => {
        mockPermission.mockReturnValueOnce({
            ...mockModule,
            default: true
        });

        const element = createIsolatedElement('c-misc-permission-based-u-i', {}, miscPermissionBasedUIModulePath);

        document.body.appendChild(element);

        // Wait for any asynchronous DOM updates
        await flushPromises();

        const pEl = element.shadowRoot.querySelector('p');
        expect(pEl.textContent).toBe('The permission set is assigned');
    });
});
```

Reply:
Hi @filiprafalowicz, Here is the working code of the isolated module with the async function. Can you please check this and make your commit. Thanks :)
`import { createElement } from 'lwc';
import MiscPermissionBasedUI from 'c/miscPermissionBasedUI';
// Mocking custom permission module
const mockModule = {
    __esModule: true
};
const mockPermission = jest.fn();
jest.mock(
    '@salesforce/customPermission/accessRestrictedUIPermission',
    () => mockPermission(),
    { virtual: true }
);
describe('c-misc-permission-based-u-i', () => {
    afterEach(() => {
        // The jsdom instance is shared across test cases in a single file so reset the DOM
        while (document.body.firstChild) {
            document.body.removeChild(document.body.firstChild);
        }
        jest.clearAllMocks();
    });
    // Helper function to wait until the microtask queue is empty. This is needed for promise
    // timing when calling imperative Apex.
    async function flushPromises() {
        return Promise.resolve();
    }
    it('displays the correct UI when custom permission is true', async () => {
        mockPermission.mockReturnValueOnce({
            ...mockModule,
            default: true
        });
        const element = createElement('c-misc-permission-based-u-i', {
            is: MiscPermissionBasedUI
        });
        document.body.appendChild(element);
        // Wait for any asynchronous DOM updates
        await flushPromises();
        const pEl = element.shadowRoot.querySelector('p');
        expect(pEl.textContent).toBe('The permission set is assigned');
    });
    it('displays the correct UI when custom permission is undefined', async () => {
        mockPermission.mockReturnValueOnce({
            ...mockModule,
            default: undefined
        });
        jest.isolateModules(async () => {
            const element = createElement('c-misc-permission-based-u-i', {
                is: MiscPermissionBasedUI
            });
            document.body.appendChild(element);
            // Wait for any asynchronous DOM updates
            await flushPromises();
            const pEl = element.shadowRoot.querySelector('p');
            expect(pEl.textContent).toBe('The permission set is not assigned');
        });
    });
    it('displays the correct UI when custom permission is false', async () => {
        mockPermission.mockReturnValueOnce({
            ...mockModule,
            default: false
        });
        jest.isolateModules(async () => {
            const element = createElement('c-misc-permission-based-u-i', {
                is: MiscPermissionBasedUI
            });
            document.body.appendChild(element);
            // Wait for any asynchronous DOM updates
            await flushPromises();
            const pEl = element.shadowRoot.querySelector('p');
            expect(pEl.textContent).toBe('The permission set is not assigned');
        });
    });
});
`

Reply:
@satyasekharcvb if you pass async function to jest.isolateModules (that doesn't accept async functions and doesn't call this function using await) tests will be always passing for you, even if they shouldn't. If you run tests from your comment above and verify the output, you will see that assertions failed but test passed. See screenshot below:
<img width="1226" alt="Zrzut ekranu 2021-08-15 o 18 34 01" src="https://user-images.githubusercontent.com/24355557/129485585-e753bd20-99a5-4f34-8e89-59a9df7cec1f.png">

You can even run test like this:
```
describe('my-test', () => {
    it('should fail when comparing true and false', async () => {
        jest.isolateModules(async () => {
            expect(false).toBe(true);
        });
    });
});
```

And you'll see it as passed.

Reply:
Yes @filiprafalowicz, it is true. I didn't realize that it is ignoring exceptions. Can you please commit the working code with the refactoring that you have done? Thanks!

Reply:
Done

Reply:
Thank you! Is it running on your system? I am getting this error when I run this code.
![Screen Shot 2021-08-16 at 4 45 51 PM](https://user-images.githubusercontent.com/31529847/129649849-b4e05967-7ae4-4661-b332-090b95445260.png)

Reply:
Yes, it does. Not sure why it is not running correctly on your machine. 
<img width="676" alt="Zrzut ekranu 2021-08-17 o 09 56 10" src="https://user-images.githubusercontent.com/24355557/129686900-5cc8984b-99e6-400b-9366-8c73f68a9473.png">